workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/RectProvider.kt

```diff
@@ -0,0 +1,37 @@
+package com.squareup.workflow1.ui.container
+
+import android.graphics.Rect
+import android.view.View
+
+/**
+ * Defines the desired bounds for a [Dialog][android.app.Dialog] built by
+ * [OverlayDialogFactory.buildDialog]. Custom containers can define their
+ * bounds policy via [DialogStack.createRectProvider].
+ *
+ * Use [ViewRectProvider] to monitor the bounds of a [View].
+ */
+public interface RectProvider {
```

Comment:
Not disagreeing but what about this:
- a dialog might receive more than one rect:
  - coverarea (what to scrim)
  - anchor view (what to appear next to... like a dropdown)
  - the proper dialog area (which can be defined by the type of dialog, for example a bottom sheet might appear horizontally centered –or using all width– and to the bottom of the cover area.

All depends where you want to place the "this type of dialog should appear here according to cover area and anchor area". If that's an "OverlayRunner" concern (as it is for our in-house library) then the runner would receive the cover-area, anchor-area, etc, rects and on each updated value recalculate the final position for the dialog (and the scrim).

Reply:
I can see the container providing the cover area -- that's what `BodyAndModalsContainer` does in this PR. But if the dialog is deciding its type, and therefore its size, it's also deciding its scrim area: cover area - body area == scrim area. Making the container responsible for both seems like it makes things much harder to customize.

It also makes sense that the container would provide the anchor view. I think the mechanism here accomodates both of those cases. I meant to imply that with this example from the other day:

```kotlin
BodyAndModals(
  body = 
    BodyAndPopUp(
      body = HomeScreen(),
      popUp = TutorialPrompt(
        targetId = "magic tutorial tag value",
        content = NowClickChargeStep()
      )
    ),
  modals: listOf<Overlay>(
    AlertModal("Oh Well", "Guess you screwed up", listOf(AlertButton("I sure did"))),
    CardModal(BackStackScreen(StepOne(), StepTwo()),
    SheetModal(SomethingThatCoversTheBody)
  )
)
```

- `BodyAndModals`, in this PR, is a container that sets `RectProvider` based on its own bounds.
  - `AlertModal` ignores RectProvider and puts up an `AlertDialog`
  - `CardModal` does the center-the-modal-in-a-scrim thing
    - Today's implementation makes a dialog that covers the entire `RectProvider` area with a transparent scrim, and centers the `BackStackScreen` within it.
    - Perhaps it would be simpler if `BodyAndModalsContainer` did its own scrim effect, and `CardAndModal` just made a centered dialog.
  - `SheetModal` fills the entire `RectProvider`
- `BodyAndPopup` would use a new container, also a built around `DialogStack`, that sets `RectProvider` to the bounds of an anchor view
  - `TutorialPrompt` understands this convention, and interprets it appropriately.
  - Probably `TutorialPrompt` implements `interface PopUp(val: idOfAnchorView): Overlay`, and `BodyAndPopup.popUp` is of type `PopUp`

That seems like it covers all of your use cases. What am I missing? Do we ever really need both a cover area and an anchor area?

> This seems to say: the creator of a dialog runner will say where it should appear

So I guess what I'm actually proposing is that the owner of a dialog can identify a rectangle of interest. Different containers can have different conventions for what that means, honored by particular dialog factories / runners.

Reply:
As we discuss this, let's remember that our internal API can still be in the shape it is now, so long as it can be driven from this with a minimum of fuss. Maybe `DialogRunner` as I'm using it here is just a different thing than your class of that name that I'm cribbing from.